File: backend/core/ingest_jobs.py
"""Shared background ingest-job registry for the storage/drive integrations.

One phantom-500 root cause, one general fix: ingesting a file or a folder
tree takes minutes (download + parse + chunk + embed + per-chunk GraphRAG),
while the Next.js dev proxy hard-aborts proxied requests at 30s
(proxy-request.js ``proxyTimeout: 30000``) and synthesizes a 500 to the
browser ("Failed to proxy ... socket hang up") while the backend is still
working. Synchronous ingest endpoints therefore surfaced as phantom 500s on
every drive integration, not just one.

The general mechanism (first built for zoho-workdrive, 2026-09-04):

- Each integration's ingest POST starts a job in THIS registry and returns
  its id immediately; heavy work runs in a request-background task.
- A repeat POST for the same file/folder while its job is running is
  COALESCED into the running job (double clicks used to race parallel tree
  walks and double the quota-limited API load).
- Each integration's router mounts the shared read routes via
  ``register_ingest_job_routes(router, integration_id)``:
    GET /ingest/jobs              recent jobs for the current user
    GET /ingest/jobs/{job_id}     job snapshot
    GET /ingest-folder/jobs/{job_id}  alias (the zoho panel's poll path)
- Durable "already ingested" badges: ``register_ingested_ids_route(router,
  source)`` exposes POST /ingested-ids, probing the document store under the
  write identity key (ext_sha1(source:external_id), incl. ::c0 chunks).

Jobs are in-process: a backend restart loses them and pollers get a 404 the
UI reports as "interrupted (the server restarted)".
"""

import asyncio
import logging
import uuid
from datetime import datetime, timezone
from typing import Any, Callable, Dict, List, Optional

from fastapi import Depends, HTTPException
from pydantic import BaseModel, Field

from core.auth import get_current_user, User

logger = logging.getLogger(__name__)

registry: Dict[str, Dict[str, Any]] = {}

_FINISHED_CAP = 50   # bound the registry: oldest finished jobs are dropped
_LIST_CAP = 20       # per-user cap on the recent-jobs list


def _now() -> str:
    return datetime.now(timezone.utc).isoformat()


def ids_key(ids: List[str]) -> str:
    return ",".join(sorted(ids))


def _prune() -> None:
    finished = sorted(
        (j for j in registry.values() if j["status"] != "running"),
        key=lambda j: j.get("finished_at") or "",
    )
    for old in finished[:-_FINISHED_CAP]:
        registry.pop(old["job_id"], None)


def create_job(
    integration_id: str,
    user_id: str,
    kind: str,
    ids: List[str],
    extra: Optional[Dict[str, Any]] = None,
) -> Dict[str, Any]:
    job_id = uuid.uuid4().hex[:12]
    job: Dict[str, Any] = {
        "job_id": job_id,
        "integration": integration_id,
        "user_id": user_id,
        "status": "running",
        "kind": kind,
        "ids_key": ids_key(ids),
        "started_at": _now(),
        "finished_at": None,
        "result": None,
        "error": None,
    }
    job.update(extra or {})
    registry[job_id] = job
    return job


def find_running(
    integration_id: str, user_id: str, kind: str, ids: List[str]
) -> Optional[Dict[str, Any]]:
    """One ingestion of a given folder/file at a time: a second POST for the
    same target while its job runs must coalesce, not race it."""
    key = ids_key(ids)
    for job in registry.values():
        if (job["status"] == "running"
                and job.get("integration") == integration_id
                and job.get("user_id") == user_id
                and job.get("kind") == kind
                and job.get("ids_key") == key):
            return job
    return None


def start_job(job: Dict[str, Any], runner: Callable[[], Any]) -> None:
    """Run ``runner`` (zero-arg async callable returning the result dict) in
    the request background; its ``success`` flag decides completed/failed.
    Feedback recording stays with the caller's runner."""
    async def _run():
        try:
            result = await runner()
            job["status"] = "completed" if result.get("success") else "failed"
            job["result"] = result
        except Exception as e:  # noqa: BLE001 — failures land in the job record
            logger.error(f"Ingest job {job['job_id']} ({job.get('integration')}) failed: {e}")
            job["status"] = "failed"
            job["error"] = str(e)[:200]
        finally:
            job["finished_at"] = _now()
            _prune()

    asyncio.create_task(_run())


def snapshot(job: Dict[str, Any]) -> Dict[str, Any]:
    return {
        "job_id": job["job_id"],
        "status": job["status"],
        "integration": job.get("integration"),
        "kind": job.get("kind"),
        "folder_ids": job.get("folder_ids"),
        "file_id": job.get("file_id"),
        "started_at": job.get("started_at"),
        "finished_at": job.get("finished_at"),
        "result": job.get("result"),
        "error": job.get("error"),
    }


def get_owned(job_id: str, integration_id: str, user_id: str) -> Optional[Dict[str, Any]]:
    job = registry.get(job_id)
    if job is None or job.get("user_id") != user_id:
        return None
    if integration_id and job.get("integration") != integration_id:
        return None
    return job


def list_for(integration_id: str, user_id: str, cap: int = _LIST_CAP) -> List[Dict[str, Any]]:
    mine = [j for j in registry.values()
            if j.get("user_id") == user_id
            and (not integration_id or j.get("integration") == integration_id)]
    running = [j for j in mine if j["status"] == "running"]
    finished = sorted(
        (j for j in mine if j["status"] != "running"),
        key=lambda j: j.get("finished_at") or j.get("started_at") or "",
        reverse=True,
    )
    return [snapshot(j) for j in (running + finished)[:cap]]


def started_payload(job: Dict[str, Any], coalesced: bool = False, **extra: Any) -> Dict[str, Any]:
    """Standard POST response for a job-backed ingest endpoint. The UI parses
    ``job_id`` from ``data`` (or top level for older payloads)."""
    data: Dict[str, Any] = {"job_id": job["job_id"], "status": "started", **extra}
    if coalesced:
        data["coalesced"] = True
    return {"success": True, "data": data,
            "message": "Ingestion already running — poll the jobs endpoint"
            if coalesced else "Ingestion started — poll the jobs endpoint"}


def register_ingest_job_routes(router: Any, integration_id: str) -> None:
    """Mount the shared read routes on an integration's router. Each router's
    prefix scopes the paths (e.g. /api/onedrive/ingest/jobs)."""

    @router.get("/ingest/jobs", summary=f"Recent {integration_id} ingestion jobs (running first)")
    async def _ingest_jobs_list(current_user: User = Depends(get_current_user)):
        return {"success": True, "data": list_for(integration_id, str(current_user.id))}

    @router.get("/ingest/jobs/{job_id}", summary="Status of an ingestion job")
    @router.get("/ingest-folder/jobs/{job_id}", summary="Status of a folder ingestion job (alias)")
    async def _ingest_job_status(job_id: str, current_user: User = Depends(get_current_user)):
        job = get_owned(job_id, integration_id, str(current_user.id))
        if job is None:
            raise HTTPException(status_code=404, detail=f"Ingestion job not found: {job_id}")
        return {"success": True, "data": snapshot(job)}


class IngestedIdsRequest(BaseModel):
    file_ids: List[str] = Field(..., min_length=1, max_length=2000,
                                description="Source-native file IDs to check against ATOM memory")


def register_ingested_ids_route(router: Any, source: str) -> None:
    """Mount POST /ingested-ids: which of these source-native file ids already
    have documents in ATOM memory. Durable source of truth for the panels'
    "already ingested" badges (session-only React state resets on reload)."""

    @router.post("/ingested-ids", summary=f"Which {source} files are already in ATOM memory")
    async def _ingested_ids(body: IngestedIdsRequest, current_user: User = Depends(get_current_user)):
        from core.auto_document_ingestion import AutoDocumentIngestionService

        ingestor = AutoDocumentIngestionService()
        found = await ingestor.ingested_external_ids(source, body.file_ids)
        return {"success": True, "data": {"ingested": found}}

File: frontend-nextjs/lib/ingest-jobs.ts
// Shared client helpers for the drive integrations' background ingest jobs.
//
// Every storage integration's ingest endpoints (zoho-workdrive, onedrive,
// gdrive, dropbox, box) start a job server-side and return
// {success, data:{job_id}} immediately — a big file or folder tree takes
// minutes to download + parse + embed, while the Next.js dev proxy aborts
// proxied requests at 30s and synthesizes a 500 ("socket hang up") while the
// backend is still working. The panels poll the job instead of holding one
// request open, and surface the shared recent-jobs list so an ingest
// started before this page load is visible instead of silently reset.

export interface IngestJob {
  job_id: string;
  status: 'running' | 'completed' | 'failed';
  kind?: string;
  integration?: string;
  folder_ids?: string[] | null;
  file_id?: string | null;
  started_at?: string | null;
  finished_at?: string | null;
  result?: any;
  error?: string | null;
}

export type FetchLike = (url: string, init?: RequestInit) => Promise<Response>;

const JOB_POLL_INTERVAL_MS = 3000;
const JOB_POLL_CAP_MS = 20 * 60 * 1000; // 20 min — folder trees are slow
const PERSISTENT_MISSING_LIMIT = 3; // consecutive 404s ⇒ the backend restarted

/**
 * POST the ingest endpoint and return the started job id (null when the
 * backend answered with a synchronous, pre-job result — backward compat).
 * Throws on HTTP errors with the server's message.
 */
export async function startIngestJob(
  fetchFn: FetchLike,
  postPath: string,
  body: Record<string, unknown>
): Promise<{ jobId: string | null; started: any }> {
  const response = await fetchFn(postPath, {
    method: 'POST',
    body: JSON.stringify(body)
  });
  if (!response.ok) {
    const text = await response.text();
    throw new Error(extractErrorMessage(text, response.status));
  }
  const started = await response.json();
  return { jobId: started?.job_id ?? started?.data?.job_id ?? null, started };
}

/**
 * Poll a job until completed/failed, then resolve with its result (already
 * normalized to {success, error?, ...}). `basePath` is the integration's API
 * base (e.g. "/api/onedrive") — every integration's router mounts the shared
 * status route at {basePath}/ingest/jobs/{job_id}. Persistent 404s mean the
 * in-process registry is gone (backend restarted) — say so instead of
 * timing out.
 */
export async function pollIngestJob(
  fetchFn: FetchLike,
  jobId: string,
  basePath: string,
  label: string
): Promise<any> {
  const deadline = Date.now() + JOB_POLL_CAP_MS;
  let consecutiveMissing = 0;
  for (;;) {
    const statusResp = await fetchFn(`${basePath}/ingest/jobs/${jobId}`);
    if (statusResp.status === 404) {
      consecutiveMissing += 1;
      if (consecutiveMissing >= PERSISTENT_MISSING_LIMIT) {
        throw new Error(
          `Ingestion job was interrupted (the server restarted) — please run the ${label} again.`
        );
      }
    } else if (statusResp.ok) {
      consecutiveMissing = 0;
      const snap = await statusResp.json();
      const job = snap?.data ?? snap;
      if (job?.status === 'completed' || job?.status === 'failed') {
        const result = job.result ?? {};
        if (job.status === 'failed') {
          return {
            ...result,
            success: false,
            error: job.error || result?.error || `${label} failed`
          };
        }
        return result;
      }
    }
    if (Date.now() > deadline) break;
    await new Promise(resolve => setTimeout(resolve, JOB_POLL_INTERVAL_MS));
  }
  throw new Error('Ingestion is still running — check the ingestion status in a few minutes.');
}

/** Convenience: start + poll in one call. `basePath` = integration base
 * (e.g. "/api/gdrive"); `postPath` = the ingest POST endpoint. */
export async function runIngestJob(
  fetchFn: FetchLike,
  postPath: string,
  basePath: string,
  body: Record<string, unknown>,
  label: string
): Promise<any> {
  const { jobId, started } = await startIngestJob(fetchFn, postPath, body);
  if (!jobId) return started; // backward compat: synchronous result
  return pollIngestJob(fetchFn, jobId, basePath, label);
}

/** Recent jobs for one integration (running first) — GET {basePath}/ingest/jobs. */
export async function fetchRecentJobs(fetchFn: FetchLike, basePath: string): Promise<IngestJob[]> {
  try {
    const response = await fetchFn(`${basePath}/ingest/jobs`);
    if (!response.ok) return [];
    const data = await response.json();
    const jobs = data?.data ?? data;
    return Array.isArray(jobs) ? jobs : [];
  } catch {
    return []; // the status strip is best-effort
  }
}

// The backend error envelope is {detail:{error:{message,...}}} (BaseAPIRouter)
// or plain {detail} / {error} — extract something human for toasts.
export function extractErrorMessage(text: string, status: number): string {
  try {
    const parsed = JSON.parse(text);
    return (
      parsed?.detail?.error?.message ||
      parsed?.detail?.message ||
      (typeof parsed?.detail === 'string' ? parsed.detail : undefined) ||
      parsed?.error ||
      `Request failed (${status})`
    );
  } catch {
    return text || `Request failed (${status})`;
  }
}
